### PR TITLE
feat(sns): add grantSubscribe method

### LINKED ISSUE #29049: (module name): (short issue description)
### Describe the feature

Add grant_subscribe method to Sns topic construct

### Use Case

Allow the Topic construct to expose a method to grant subscription permissions to a grantable resource. Would make it simpler instead of having to add the Policy statement to the resource policy. 

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.126.0

### Environment details (OS name and version, etc.)

macOS 14.2.1

### PR DESCRIPTION
**Pull request by mazyu36**
### Issue # (if applicable)

Closes #29049.

### Reason for this change
Allow the Topic construct to expose a method to grant subscription permissions to a grantable resource.
It's useful when you want to allow entities, such as another AWS account or resources created later, to subscribe to the topic at their own pace, separating permission granting from the actual subscription process.

### Description of changes
Add grantSubscribe method to ITopic interface and TopicBase class.

### Description of how you validated changes
Add unit tests and integ tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-sns/README.md [hunk 1]
```diff
@@ -61,6 +61,16 @@ myTopic.addSubscription(new subscriptions.SqsSubscription(queue));
 Note that subscriptions of queues in different accounts need to be manually confirmed by
 reading the initial message from the queue and visiting the link found in it.
 
+ The `grantSubscribe` method adds a policy statement to the topic's resource policy, allowing the specified principal  to perform the `sns:Subscribe` action.
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-sns/README.md:64]:
```suggestion
 The `grantSubscribe` method adds a policy statement to the topic's resource policy, allowing the specified principal to perform the `sns:Subscribe` action.
```

**Review by lpizzinidev — APPROVED** [review]
Thanks 👍

**Review by aaythapa — APPROVED** [review]
lgtm

**Comment by CaptainDriftwood** [comment]
Thanks 🚀

**Pull request merged**